app/services/foreman_discovery/host_converter.rb

```diff
@@ -21,9 +21,9 @@ def self.to_managed(original_host, set_managed = true, set_build = true, added_a
   def self.set_build_clean_facts(host)
     # fact cleaning
     if Setting['discovery_clean_facts']
-      # clean all facts except those starting with "discovery_"
+      # clean all facts except those from Discovery
       host.define_singleton_method(:clear_facts) do
-          keep_ids = FactValue.where("host_id = #{host.id}").joins(:fact_name).where("fact_names.name like 'discovery_%'").pluck("fact_values.id")
+          keep_ids = FactValue.where("host_id = #{host.id}").joins(:fact_name).where("fact_names.type like 'DiscoveryFactName'").pluck("fact_values.id")
           FactValue.where("host_id = #{host.id} and id not in (?)", keep_ids).delete_all
```

Comment:
This could be refactored to use pure rails notation.
```ruby
FactValue.where.not(id: keep_ids)
```
Even better would be to use subqueries.

Reply:
Thanks for the idea! Select is refactored.